payu/models/staged_cable.py

```diff
@@ -0,0 +1,205 @@
+            config['input'].append(os.path.realpath(
+                f'archive/output{restart_dir:03d}/restart'))
+
+        # Now set the number of runs using the configuration_log
+        remaining_stages = len(self.configuration_log['queued_stages'])
+        os.environ['PAYU_N_RUNS'] = str(remaining_stages)
+
+    def _build_new_configuration_log(self):
+        """Build a new configuration log for the first stage of the run."""
+
+        # On the first run, we need to read the 'stage_config.yaml' file.
+        cable_stages = self._prepare_configuration()
+
+        # Build the configuration log
+        self.configuration_log['queued_stages'] = cable_stages
+        self.configuration_log['current_stage'] = ''
+        self.configuration_log['completed_stages'] = []
+
+        with open('configuration_log.yaml', 'w') as conf_log_file:
+            yaml.dump(self.configuration_log, conf_log_file)
+
+    def _read_configuration_log(self):
+        """Read the existing configuration log."""
+        with open('configuration_log.yaml') as conf_log_file:
+            self.configuration_log = yaml.safe_load(conf_log_file)
+
+    def _prepare_configuration(self):
+        """Prepare the stages in the CABLE configuration."""
+
+        # Since Python3.7, dictionary order is guaranteed so we can read
+        # the entries in order without needing to supply an index
+        # We just want to populate cable_stages with the list of stages
+        # to run
+        cable_stages = []
+        for stage_name, stage_opts in self.stage_config.items():
+            # Check if stage is a multi-step or single step
+            if stage_name.startswith('multistep'):
+                # The multi-step stage can run each internal stage
+                # a different number of times. For example, a two
+                # step stage may ask for the first step (S1) 5 times,
+                # but the second step (S2) only 3 times. The stage
+                # looks like [S1, S2, S1, S2, S1, S2, S1, S1].
+                # So what we need to do is first record the number
+                # of times each step is run
+                step_names = []
+                step_counts = []
+                for step_name, step_opts in stage_opts.items():
+                    step_names.append(step_name)
+                    step_counts.append(step_opts['count'])
+
+                # Now iterate to the maximum number of steps
+                for step in range(max(step_counts)):
+                    for step_id, _ in enumerate(step_names):
+                        if step_counts[step_id] > step:
+                            cable_stages.append(step_names[step_id])
+            # Finish handling of multistep stage
+
+            else:
+                # A single step stage, in general we only want to run this
+                # once, but check for the count anyway
+                for _ in range(stage_opts['count']):
+                    cable_stages.append(stage_name)
```

Comment:
I think this can be replaced with something like:
```python
for stage, repeat in self.stage_config.items();
   cable_stages.extend([stage] * repeat))
```
Possibly could be done as a one-liner list comprehension, but I got a bit bogged down in the syntax, and this is relatively clean and understandable.

Reply:
That will work for the single step stages, but I don't think it will for the multi-step stages. An example of a multistep stage in the config file would be
```
multistep_demo_stage:
    demo_stage_1:
        count: 4
    demo_stage_2:
        count: 2
```
The desired output is ```["demo_stage_1, "demo_stage_2", "demo_stage_1", "demo_stage_2", "demo_stage_1", "demo_stage_1"]```, but I think the one liner would give ```["demo_stage_1", "demo_stage_1", "demo_stage_1", "demo_stage_1", "demo_stage_2", "demo_stage_2"]```. There is perhaps a cleaner way to do it than the way I did it, but I couldn't think of it at the time.

Reply:
Ah, sorry my misunderstanding. This is where tests are super-useful (some might even say essential), as it gives a very clear indication of what the expected output would be. 

Depending on what is required [itertools](https://docs.python.org/3/library/itertools.html) provides a bunch of useful functions for these sorts of shenanigans.

Reply:
Is cutting down the number of lines so important? I think it's not a such a bad idea to have it explicitly written with simple methods, but then again, the implementation didn't make clear what was going on. Perhaps a better idea is to find a convenient method and make the explanation in the comments clearer.

Reply:
I agree with @aidanheerdegen that I'm not sure what is going on in the code here. What about using `itertools.zip_longest()`? You pad the shortest stages with a given missing value, then you can use list comprehension to remove this missing value from the list.

This has the advantage that people know what `zip` does in Python.

Reply:
Agreed clarity is the most important factor, and I didn't read the code carefully enough to understand what it was trying to do which was poor, sorry. But as you say, perhaps it could be expressed more clearly. Also tests. Did I mention tests?

Using something from `itertools` at least means using a well documented function, with a widely understood purpose.

Adapted from https://stackoverflow.com/a/48200012 this shows one way of doing it.
```python
>>> list(filter( lambda x: x is not None, chain.from_iterable(zip_longest(['a', 'b', 'c', 'd'], [1, 2]))))
['a', 1, 'b', 2, 'c', 'd']
```

By itself maybe not that readable, but as you say with a comment line "Build a list interleaving steps using `zip_longest`, and use `chain` and `filter` to flatten and remove missing elements"

Or you could use a [recipe from the itertools docs](https://docs.python.org/3/library/itertools.html#itertools-recipes):
```python
def roundrobin(*iterables):
    "Visit input iterables in a cycle until each is exhausted."
    # roundrobin('ABC', 'D', 'EF') → A D E B F C
    # Algorithm credited to George Sakkis
    iterators = map(iter, iterables)
    for num_active in range(len(iterables), 0, -1):
        iterators = cycle(islice(iterators, num_active))
        yield from map(next, iterators)
```

Reply:
I have adapted the first suggestion with some minor changes, compacting the handling of the multi-step stage to a few lines.

Reply:
Weird we overlapped with identical suggestions @ccarouge. I think I must have started my comment before yours, so didn't see it, otherwise I clearly wouldn't have bothered. Yours was much more succinct! :)

Reply:
Ended up going with an adaptation of the first solution.